### PR TITLE
Use uuid.uuid4() as the key in serialization.

### LINKED ISSUE #7881: objmode() triggers fork()
## Reporting a bug

- [x] I have tried using the latest released version of Numba (most recent is
 visible in the change log (https://github.com/numba/numba/blob/main/CHANGE_LOG).
- [x] I have included a self contained code sample to reproduce the problem.
  i.e. it's possible to run as 'python bug.py'.

It seems that using [`with numba.objmode()`](https://numba.readthedocs.io/en/stable/user/withobjmode.html) is triggering a call to `fork()` somewhere. In an MPI context, this can be problematic – Open MPI is complaining about this with a message of the following form:

>     --------------------------------------------------------------------------
>     A process has executed an operation involving a call to the
>     "fork()" system call to create a child process.  Open MPI is currently
>     operating in a condition that could result in memory corruption or
>     other system errors; your job may hang, crash, or produce silent
>     data corruption.  The use of fork() (or system() or other calls that
>     create child processes) is strongly discouraged.
>     
>     The process that invoked fork was:
>     
>       Local host:          [[53156,1],2] (PID 26851)
>     
>     If you are *absolutely sure* that your application will successfully
>     and correctly survive a call to fork(), you may disable this warning
>     by setting the mpi_warn_on_fork MCA parameter to 0.
>     --------------------------------------------------------------------------

Using the same code, but omitting `with numba.objmode()`, Open MPI doesn’t complain. As far as I can remember, this issue only started to appear recently, so there seems to have been a change in the implementation causing this? If there is some way to stop `objmode()` from triggering a `fork()`, that would be great.

<details>
  <summary>(Trivial) code sample</summary>

```
import numba
from mpi4py import MPI

mpi_comm = MPI.COMM_WORLD
mpi_rank = mpi_comm.Get_rank()

@numba.njit
def f():
    with numba.objmode():
        print(mpi_rank)

f()
```
</details>

<details>
  <summary>Running the sample and output</summary>

>     $ mpiexec -n 3 python3 test.py
>     --------------------------------------------------------------------------
>     A process has executed an operation involving a call to the
>     "fork()" system call to create a child process.  Open MPI is currently
>     operating in a condition that could result in memory corruption or
>     other system errors; your job may hang, crash, or produce silent
>     data corruption.  The use of fork() (or system() or other calls that
>     create child processes) is strongly discouraged.
>     
>     The process that invoked fork was:
>     
>       Local host:          [[53156,1],2] (PID 26851)
>     
>     If you are *absolutely sure* that your application will successfully
>     and correctly survive a call to fork(), you may disable this warning
>     by setting the mpi_warn_on_fork MCA parameter to 0.
>     --------------------------------------------------------------------------
>     1
>     2
>     0
>     [host:26836] 2 more processes have sent help message help-opal-runtime.txt / opal_init:warn-fork
>     [host:26836] Set MCA parameter "orte_base_help_aggregate" to 0 to see all help / error messages
</details>

### PR DESCRIPTION
**Pull request by stuartarchibald**
As title, this is to avoid fork()+exec() calls made by uuid1().

Fixes #7881

### CODE REVIEW

File: numba/tests/test_withlifting.py [hunk 1]
```diff
@@ -1213,5 +1220,30 @@ def foo():
...
 
 
+class TestMisc(TestCase):
+    # Tests for miscellaneous objmode issues. Run serially.
+
+    _numba_parallel_test_ = False
+
+    @linux_only
+    @needs_strace
+    def test_no_fork_in_compilation(self):
```
Comment (stuartarchibald) [inline numba/tests/test_withlifting.py:1230]: Note to self. This test probably needs to be run as a subprocess test on the basis that previous execution of something which triggers a stateful call e.g. `uuid.uuid1.getnode()` is pretty likely.

Comment (stuartarchibald) [reply numba/tests/test_withlifting.py]: Addressed in d114552

File: numba/tests/support.py [hunk 2]
```diff
@@ -1007,6 +1009,98 @@ def run_in_subprocess(code, flags=None, env=None, timeout=30):
...
+                check_return(problem)
+            # Make sure the return code is 0, SIGINT to detach is considered
+            # a successful exit.
+            popen.wait()
+            check_return()
+            # collect the data
+            strace_data = ntf.readlines()
+            # stop the process
+            popen.terminate()
+            if popen.poll() is None:
```
Comment (sklam) [inline numba/tests/support.py:1069]: Would there ever be a delay in `terminate()` that `poll()` will see the process not terminated but actually in the process of doing so? Should this be a `wait()`?

File: numba/tests/support.py [hunk 3]
```diff
@@ -1007,6 +1009,98 @@ def run_in_subprocess(code, flags=None, env=None, timeout=30):
...
+
+def _strace_supported():
+    """Checks if strace is supported and working"""
+    def force_clone(): # subprocess triggers a clone
+        subprocess.run([sys.executable, '-c', 'exit()'])
+
+    try:
+        syscall = 'clone'
+        trace = strace(force_clone, [syscall,])
+        assert syscall in ''.join(trace)
```
Comment (sklam) [inline numba/tests/support.py:1091]: not why just `return`?

File: numba/tests/support.py [hunk 4]
```diff
@@ -1007,6 +1009,98 @@ def run_in_subprocess(code, flags=None, env=None, timeout=30):
...
+
+
+def _strace_supported():
+    """Checks if strace is supported and working"""
+    def force_clone(): # subprocess triggers a clone
+        subprocess.run([sys.executable, '-c', 'exit()'])
+
+    try:
+        syscall = 'clone'
+        trace = strace(force_clone, [syscall,])
```
Comment (sklam) [inline numba/tests/support.py:1090]: try..except should only be applied to this line

**Review by sklam — COMMENTED** [review]
Thanks for the patch! `uuid4` should be sufficient given its use in cloudpickle and its implementation should never depend on fork (hopefully).

There are a few comments esp. in the strace code regarding the handling of the subprocess

File: numba/tests/support.py (hunk 2 above)
Comment (sklam) [reply numba/tests/support.py:1069]: right, i missed the earlier `wait()`.  and yes for the removal

File: numba/tests/support.py (hunk 3 above)
Comment (stuartarchibald) [reply numba/tests/support.py:1091]: Ah I see, yes, thanks, will fix.

**Review by sklam — APPROVED** [review]
Thanks for the patch!

**Pull request merged**